### PR TITLE
ISSUE-472 implement buildUrlPattern function to add url matching regardless query params

### PR DESCRIPTION
**Pull request by MrXavier**
# ISSUE-472 implement buildUrlPattern function to add url matching regardless query params

PR Branch
https://github.com/MrXavier/zerocode/tree/ISSUE-472_wiremock_match_url_path_only

## Motivation and Context
Scenario thi PR wants to cover:
Url used in mock step scenario = /google-guys/persons/p001
Request send to Wiremock server = /google-guys/persons/p001?param1=value1&param2=value2
Result = The request is mocked and return the mocked response as expected, regardless the query params.

Again, this suggested change is to allow mock non deterministic requests with random query parameters, like for example, when a request has ID or UUID as query parameter. In this case the values of query params can be any random value. To be able to mock that, Wiremock provides urlPathEqualTo(url) function to solve this exact scenario, which is used in this PR to improve mocking feature in Zerocode.

All described here https://github.com/authorjapps/zerocode/issues/472

## Checklist:

* [ ] Unit tests added

* [ ] Integration tests added

* [ ] Test names are meaningful

* [ ] Feature manually tested

* [ ] Branch build passed

* [ ] No 'package.*' in the imports

* [ ] Relevant Wiki page updated with clear instruction for the end user
  * [ ] Not applicable. This was only a refactor change, no functional or behaviour changes were introduced

* [ ] Http test added to `http-testing` module(if applicable) ?
  * [ ] Not applicable. The changes did not affect HTTP automation flow

* [ ] Kafka test added to `kafka-testing` module(if applicable) ?
  * [ ] Not applicable. The changes did not affect Kafka automation flow

### CODE REVIEW

File: core/src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java [hunk 1]
```diff
@@ -95,30 +97,38 @@ public static void stopWireMockServer() {
...
         return createRequestBuilderWithHeaders(mockStep, requestBuilder);
     }
 
     private static MappingBuilder createGetRequestBuilder(MockStep mockStep) {
-        final MappingBuilder requestBuilder = get(urlEqualTo(mockStep.getUrl()));
+        final MappingBuilder requestBuilder = get(buildUrlPattern(mockStep.getUrl()));
         return createRequestBuilderWithHeaders(mockStep, requestBuilder);
     }
 
+    private static UrlPattern buildUrlPattern(String url) {
```
Comment (imprashant) [inline core/src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java:124]: Sorry I could not find tests related to this, I mean not for this private function but for the change in logic for url matching.  
Can you please point to the tests/scenarios where this strict matching vs path matching is needed

**Comment by sparrowV** [comment]
@authorjapps does docker issue that we faced last time, still persits?

File: core/src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java (hunk 1 above)
Comment (sparrowV) [reply core/src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java:124]: @MrXavier  please add appropriate tests for this one.

**Comment by sparrowV** [comment]
@MrXavier  please add appropriate tests/scenarios to make sure that feature works as expected

**Comment by MrXavier** [comment]
Hello @sparrowV and @imprashant, I added the tests as suggested and they work as expected. You can see that when mocking a url path without parameters, it accepts HTTP requests without or with any query params.
Scenario:
Url used in mock step scenario = `/google-guys/persons/p001`
Request send to Wiremock server = `/google-guys/persons/p001?param1=value1&param2=value2`
Result = The request is mocked an return the mocked response as expected.

Again this suggested change is to allow mock non deterministic requests with random query parameters, like for example, when a request has ID or UUID as query parameter. In this case the values of query params can be nay random value. To be able to mock that Wiremock provides `urlPathEqualTo(url)` function, to solve this exact scenario, and this is used in this PR to improve mocking feature in Zerocode.

**Comment by sparrowV** [comment]
@MrXavier  so now if we have some query params in url it always matches to the step,that declares the base url,right? As you wrote,if
we have `/google-guys/persons/p001?param1=value1&param2=value2` it will match `/google-guys/persons/p001` , but what if an user wants a seperate step where he can specify specific query params? For example: 
```
 "mocks": [
                    {
                        "name": "mocking_a_GET_endpoint_with_headers",
                        "operation": "GET",
                        "url": "/api/v1/amazon/customers/cust-007",
                        "request": {
                            "headers": {
                                "api_key": "key-01-01",
                                "api_secret": "secret-01-01"
                            }
                        },
                        "response": {
                            "status": 200,
                            "body": {
                                "id": "cust-007",
                                "type": "Premium"
                            },
                            "headers": {
                                "Content-Type": "application/json"
                            }
                        }
                    },{
                        "name": "mocking_a_GET_endpoint_with_headers",
                        "operation": "GET",
                        "url": "/api/v1/amazon/customers/cust-007?param1=param1",
                        "request": {
                            "headers": {
                                "api_key": "key-01-01",
                                "api_secret": "secret-01-01"
                            }
                        },
                        "response": {
                            "status": 302,
                            "body": {
                                "id": "cust-007",
                                "type": "Premium"
                            },
                            "headers": {
                                "Content-Type": "application/json"
                            }
                        }
                    }
                ]
            },
            "assertions": {
                "status": 200
            }
        },
        //------------- Mocking done at this point via WireMock ---------------

        //------------- Let's verify the end points by writing the following small tests ------------
        {
            "name": "verify_get_customer_with_headers",
            "url": "/api/v1/amazon/customers/cust-007?param1=value1",
            "operation": "GET",
            "request": {
                "headers": {
                    "api_key": "key-01-01", //<--- Please try with a wrong key. The test should fail.
                    "api_secret": "secret-01-01"
                }
            },
            "assertions": {
                "status": 200,
                "body": {
                    "id": "cust-007",
                    "type": "Premium"
                },
                "headers": {
                    "Content-Type": ["application/json"]
                }
            }
        },
```

In step `verify_get_customer_with_headers`, Im trying to match the second mock with url `"url": "/api/v1/amazon/customers/cust-007?param1=param1",`, but It matches the first one.  I think it would be better if firstly we try to match exact url, but if exact fails match base url.

**Comment by MrXavier** [comment]
Hello @sparrowV, yes exactly, when you set the url with only path, without query params, the mock will match any requests with that specific path, regardless wether there are query params or not.
In case the mock specify the url with query parameter, the behaviour will be the same as before, the mock will match only if the path and query parameters match exactly as it is defined in the mock.

The scenario you mentioned you are defining 2 mocks for the same url path, but one has query parameters. This is an edge case which might lead to conflicts inside Wiremock framework originally (note that what I am doing is basically using Wiremock functions to define the request url pattern. Any combinations of mock you use will follow the logic that Wiremock implemented internally).
In my opinion, I don't think someone should mock some url path and the same path with query parameters in the same test scenario. This have logical conflicts and undetermined behaviour by nature.

Hope this answer your question! In the end, I don't have a simple solution for the scenario you mentioned. It would be necessary to debug Wiremock internally to try to make your scenario to work.

**Comment by sparrowV** [comment]
> Hello @sparrowV, yes exactly, when you set the url with only path, without query params, the mock will match any requests with that specific path, regardless wether there are query params or not.
> In case the mock specify the url with query parameter, the behaviour will be the same as before, the mock will match only if the path and query parameters match exactly as it is defined in the mock.
> 
> The scenario you mentioned you are defining 2 mocks for the same url path, but one has query parameters. This is an edge case which might lead to conflicts inside Wiremock framework originally (note that what I am doing is basically using Wiremock functions to define the request url pattern. Any combinations of mock you use will follow the logic that Wiremock implemented internally).
> In my opinion, I don't think someone should mock some url path and the same path with query parameters in the same test scenario. This have logical conflicts and undetermined behaviour by nature.
> 
> Hope this answer your question! In the end, I don't have a simple solution for the scenario you mentioned. It would be necessary to debug Wiremock internally to try to make your scenario to work.



If you do it , that would be great, We should just check first if it URL matches strictly and then if it doesn't pick the one with base URL.  You can go for it and dig into, or maybe I will do it if I get some time :)

**Comment by MrXavier** [comment]
Hello @sparrowV, I added a feature flag and enable the build urlPathEqualTo pattern only if the url appears only once in the mocks.
This is a work around to tackle your scenario. Unfortunately, I don't have time to dig more into this solution. Feel free to try a better solution. If not I would say this PR is good to be merged to allow this important mock feature to be used in Zerocode.

**Comment by authorjapps** [comment]
Hello mate,
The branch builds are getting though temporarily. Visit this [page](https://github.com/authorjapps/zerocode/wiki/Forked-Builds-Failing-Due-to-Docker-Hub-Rate-Limiting) for the solution.

**Comment by authorjapps** [comment]
> @authorjapps does docker issue that we faced last time, still persits?

Still persists mate. Until then the solution is [here](https://github.com/authorjapps/zerocode/wiki/Forked-Builds-Failing-Due-to-Docker-Hub-Rate-Limiting).

**Comment by MrXavier** [comment]
Hello @authorjapps thank you for your comment. I didn't have time to follow the instructions you sent. Will do it shortly, though.

**Pull request closed without merging**